### PR TITLE
SE: Learn NumberConstraint from existing constraint

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Part of #7143

Commits `Squash LearnNumberCombined`  does not need a review
Coverage will be fixed by a rebase, a commit was already pushed to #7198 

Review per commit will be easier than the full batch

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 1]
```diff
@@ -92,57 +93,65 @@ private static ProgramState LearnBranchingEqualityConstraint<T>(ProgramState sta
...
 
-        static NumberConstraint RelationalNumberConstraint(BinaryOperatorKind kind, NumberConstraint constraint) =>
+        static NumberConstraint RelationalNumberConstraint(NumberConstraint existingNumber, BinaryOperatorKind kind, NumberConstraint comparedNumber) =>
             kind switch
             {
-                BinaryOperatorKind.GreaterThan when constraint.Min.HasValue => NumberConstraint.From(constraint.Min + 1, null),
-                BinaryOperatorKind.GreaterThanOrEqual when constraint.Min.HasValue => NumberConstraint.From(constraint.Min, null),
-                BinaryOperatorKind.LessThan when constraint.Max.HasValue => NumberConstraint.From(null, constraint.Max - 1),
-                BinaryOperatorKind.LessThanOrEqual when constraint.Max.HasValue => NumberConstraint.From(null, constraint.Max),
+                BinaryOperatorKind.Equals => comparedNumber,
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:123]: This breaks for cases where none of `existingNumber` and `comparedNumber` are single values. E.g., [0..100] and [42..150].

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 2]
```diff
@@ -92,57 +93,65 @@ private static ProgramState LearnBranchingEqualityConstraint<T>(ProgramState sta
...
         }
 
-        ProgramState LearnBranching(ISymbol leftSymbol, BinaryOperatorKind kind, NumberConstraint rightConstraint) =>
-            !(falseBranch && leftSymbol.GetSymbolType().IsNullableValueType())  // Don't learn opposite for "nullable > 0", because it could also be <null>.
-            && RelationalNumberConstraint(kind, rightConstraint) is { } leftConstraint
-                ? state.SetSymbolConstraint(leftSymbol, leftConstraint)
+        ProgramState LearnBranching(ISymbol symbol, NumberConstraint existingNumber, BinaryOperatorKind kind, NumberConstraint comparedNumber) =>
+            !(falseBranch && symbol.GetSymbolType().IsNullableValueType())  // Don't learn opposite for "nullable > 0", because it could also be <null>.
+            && RelationalNumberConstraint(existingNumber, kind, comparedNumber) is { } leftConstraint
+                ? state.SetSymbolConstraint(symbol, leftConstraint)
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:117]: We have `existing`, so the new one should be indeed `new`

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:123]: I'll address this in separate PR

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM

**Pull request merged**